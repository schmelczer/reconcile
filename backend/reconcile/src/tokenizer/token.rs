#[cfg(feature = "serde")]
use serde::{Deserialize, Serialize};

/// A token is a string that has been normalised in some way.
/// The normalised form is used for comparison, while the original form is used
/// for applying Operations.
#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
#[derive(Debug, Clone)]
pub struct Token<T>
where
    T: PartialEq + Clone + std::fmt::Debug,
{
    normalised: T,
    original: String,
}

impl From<&str> for Token<String> {
    fn from(s: &str) -> Self { Token::new(s.trim().to_owned(), s.to_owned()) }
}

impl<T> Token<T>
where
    T: PartialEq + Clone + std::fmt::Debug,
{
    pub fn new(normalised: T, original: String) -> Self {
        Token {
            normalised,
            original,
        }
    }

    pub fn original(&self) -> &str { &self.original }

    pub fn normalised(&self) -> &T { &self.normalised }

    pub fn get_original_length(&self) -> usize { self.original.chars().count() }
}

impl<T> PartialEq for Token<T>
where
    T: PartialEq + Clone + std::fmt::Debug,
{
    fn eq(&self, other: &Self) -> bool { self.normalised == other.normalised }
}
