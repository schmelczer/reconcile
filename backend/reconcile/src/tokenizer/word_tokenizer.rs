use super::token::Token;

/// Splits on whitespace keeping the leading whitespace.
///
///     
/// ## Example
///
/// "Hi there!" -> ["Hi", " there!"]
pub fn word_tokenizer(text: &str) -> Vec<Token<String>> {
    let mut result: Vec<Token<String>> = Vec::new();

    let mut last_whitespace = 0;
    let mut previous_char_is_whitespace = true;

    for (i, c) in text.char_indices() {
        let is_current_char_whitespace = c.is_whitespace();
        if !previous_char_is_whitespace && is_current_char_whitespace {
            result.push(text[last_whitespace..i].into());
            last_whitespace = i;
        }

        previous_char_is_whitespace = is_current_char_whitespace;
    }

    if last_whitespace < text.len() {
        result.push(text[last_whitespace..].into());
    }

    result
}

#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;

    use super::*;

    #[test]
    fn test_with_snapshots() {
        assert_debug_snapshot!(word_tokenizer("Hi there!"));

        assert_debug_snapshot!(word_tokenizer(""));

        assert_debug_snapshot!(word_tokenizer(" what? "));

        assert_debug_snapshot!(word_tokenizer(" hello, \nwhere are you?"));
    }
}
