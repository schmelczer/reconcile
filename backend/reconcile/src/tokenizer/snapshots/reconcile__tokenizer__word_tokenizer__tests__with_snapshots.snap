---
source: reconcile/src/tokenizer/word_tokenizer.rs
expression: "word_tokenizer(\"Hi there!\")"
snapshot_kind: text
---
[
    Token {
        normalised: "Hi",
        original: "Hi",
    },
    Token {
        normalised: "there!",
        original: " there!",
    },
]
